sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,605 @@
+code, not the JIT compiler that translates SPIR-V into native code.
+
+In general, the front-end compiler does not know which kernels the application
+will submit to which devices.  Therefore, the front-end compiler does not
+generally know which optional features a kernel can legally use.  Thus, in
+general, the front-end compiler must not issue any diagnostic simply because a
+kernel uses an optional feature.
+
+The only exception to this rule occurs when the application uses the C++
+attribute `[[sycl::requires()]]`.  When the application decorates a kernel or
+device function with this attribute, it is an assertion that the kernel or
+device function is allowed to use only those optional features which are listed
+by the attribute.  Therefore, the front-end compiler must issue a diagnostic if
+the kernel or device function uses any other optional kernel features.
+
+Note that this behavior does not change when the compiler runs in AOT mode.
+Even if the user specifies a target device via "-fsycl-targets", that does not
+necessarily mean that the user expects all the code in the application to be
+runnable on that device.  Consider an application that uses some middleware
+library, where the library's header contains kernels optimized for different
+devices.  An application should be able to compile in AOT mode with this
+library without getting errors.  Therefore the AOT compiler must not fail
+simply because the middleware header contains device code for devices that
+are not being compiled for.
+
+### Runtime exception if device doesn't support feature
+
+When the application submits a kernel to a device via one of the kernel
+invocation commands (e.g. `parallel_for()`), the runtime must check if the
+kernel uses optional features that are not supported on that device.  If the
+kernel uses an unsupported feature, the runtime must throw a synchronous
+`errc::kernel_not_supported` exception.
+
+This exception must be thrown in the following circumstances:
+
+* For a kernel that is not decorated with `[[sycl::requires()]]`: the exception
+  must be thrown if a device function in the kernel's static call tree uses a
+  feature that the device does not support.  However, this only applies to
+  features that are exposed via a C++ type or function.  Examples of this
+  include `sycl::half` or instantiating `sycl::atomic_ref` for a 64-bit type.
+  If the kernel relies on optional features that are more "notional" such as
+  sub-group independent forward progress
+  (`info::device::sub_group_independent_forward_progress`), no exception is
+  required.
+
+* For a kernel that is decorated with `[[sycl::requires()]]`: the exception
+  must be thrown if the device does not have the aspects listed in that
+  attribute.  Note that the exception must be thrown even if the kernel does
+  not actually use a feature corresponding to the aspect, and it must be
+  thrown even if the aspect does not correspond to any optional feature.
```

Comment:
How does the runtime know what aspects are listed in the attribute -- there's no reflection over attributes?

Reply:
Those attributes are expected to be lowered into metadata/attributes at LLVM IR level and `sycl-post-link` together with `clang-offload-wrapper` will embed them into device image properties, which are available to DPC++ RT.

Reply:
Ah, so the reflection is generated by later stages in the pipeline, good to know. Thanks!